### PR TITLE
OCPBUGS-31679: localnet, multi-homing: introduce localnet alias

### PR DESCRIPTION
**Pull request by maiqueb**
**- What this PR does and why is it needed**

This PR introduces a network name alias to use when configuring the localnet networks.

By using this alias, the user can create multiple networks - on different vlans - without having to reconfigure the node's OVN bridge mappings. Essentially, the use can provide a **single** mapping and piggy-back multiple physical configurations in it.

Users in OpenShift use kubernetes-nmstate to configure these mappings; this way, they get to skip provisioning multiple policies. 

**- Special notes for reviewers**

In this PR we are adding an alias to the network name attribute in the localnet logical switch port mandatory options.

By doing that, we can create multiple networks pointing at the same OVS bridge without having to provision additional bridge mappings, which provides a leaner and more focuses user workflow since any time the admin wants to define a new VLAN, all they need to do is to create a NetworkAttachmentDefinition (and re-use the existing binding).

This would be the new secondary networks definitions:
```yaml
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: tenantblue123
spec:
  config: |2
    {
            "cniVersion": "0.4.0",
            "name": "tenantblue123",
            "type": "ovn-k8s-cni-overlay",
            "topology":"localnet",
            "vlanID": 123,
            "physicalNetworkName": "tenantblue",
            "subnets": "192.100.200.0/24",
            "excludeSubnets": "192.100.200.1/32",
            "netAttachDefName": "default/tenantblue123"
    }
---
apiVersion: k8s.cni.cncf.io/v1
kind: NetworkAttachmentDefinition
metadata:
  name: tenantblue321
spec:
  config: |2
    {
            "cniVersion": "0.4.0",
            "name": "tenantblue321",
            "type": "ovn-k8s-cni-overlay",
            "topology":"localnet",
            "vlanID": 321,
            "physicalNetworkName": "tenantblue",
            "subnets": "192.150.250.0/24",
            "excludeSubnets": "192.150.250.1/32",
            "netAttachDefName": "default/tenantblue321"
    }
```

**- How to verify it**

Provision workloads ataching to those networks, and check the OVS databases contents:
```bash
# logical switch ports (localnet)
_uuid               : 014037c2-e34a-4ae1-b798-23133dbba72b
addresses           : [unknown]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {}
ha_chassis_group    : []
mirror_rules        : []
name                : tenantblue321_ovn_localnet_port
options             : {network_name=tenantblue}
parent_name         : []
port_security       : []
tag                 : 321
tag_request         : 321
type                : localnet
up                  : false


_uuid               : f9092ca2-a876-47c1-9897-1b12be6d573f
addresses           : [unknown]
dhcpv4_options      : []
dhcpv6_options      : []
dynamic_addresses   : []
enabled             : []
external_ids        : {}
ha_chassis_group    : []
mirror_rules        : []
name                : tenantblue123_ovn_localnet_port
options             : {network_name=tenantblue}
parent_name         : []
port_security       : []
tag                 : 123
tag_request         : 123
type                : localnet
up                  : false

# OVS bridges port configuration
kubectl exec ovs-node-m6ch6 -novn-kubernetes -- ovs-vsctl show  
27800d09-e777-4798-86c6-45a5adfc5079
    Bridge breth0
        fail_mode: standalone
        Port patch-tenantblue123_ovn_localnet_port-to-br-int
            Interface patch-tenantblue123_ovn_localnet_port-to-br-int
                type: patch
                options: {peer=patch-br-int-to-tenantblue123_ovn_localnet_port}
        Port breth0
            Interface breth0
                type: internal
        Port patch-breth0_ovn-worker-to-br-int
            Interface patch-breth0_ovn-worker-to-br-int
                type: patch
                options: {peer=patch-br-int-to-breth0_ovn-worker}
        Port patch-tenantblue321_ovn_localnet_port-to-br-int
            Interface patch-tenantblue321_ovn_localnet_port-to-br-int
                type: patch
                options: {peer=patch-br-int-to-tenantblue321_ovn_localnet_port}
        Port eth0
            Interface eth0
    Bridge br-int
        fail_mode: secure
        datapath_type: system
...
        Port patch-br-int-to-breth0_ovn-worker
            Interface patch-br-int-to-breth0_ovn-worker
                type: patch
                options: {peer=patch-breth0_ovn-worker-to-br-int}
        Port patch-br-int-to-tenantblue321_ovn_localnet_port
            Interface patch-br-int-to-tenantblue321_ovn_localnet_port
                type: patch
                options: {peer=patch-tenantblue321_ovn_localnet_port-to-br-int}
        Port patch-br-int-to-tenantblue123_ovn_localnet_port
            Interface patch-br-int-to-tenantblue123_ovn_localnet_port
                type: patch
                options: {peer=patch-tenantblue123_ovn_localnet_port-to-br-int}
        Port "43bd51029b658_3"
            Interface "43bd51029b658_3"
    ovs_version: "3.2.2"

# OVS global configuration - check the external ids
kubectl exec ovs-node-m6ch6 -novn-kubernetes -- ovs-vsctl list open .
_uuid               : 27800d09-e777-4798-86c6-45a5adfc5079
bridges             : [580eae59-db86-4eb8-81f1-da44f4a0a23d, e65ed98f-bc0d-46d3-aef6-bf67aa67e517]
cur_cfg             : 45
datapath_types      : [netdev, system]
datapaths           : {system=a7910ab8-fe2d-4df8-83bd-b889d684fd7d}
db_version          : "8.4.0"
dpdk_initialized    : false
dpdk_version        : "DPDK 22.11.1"
external_ids        : {hostname=ovn-worker, ovn-bridge-mappings="physnet:breth0,tenantblue:breth0", ovn-enable-lflow-cache="true", ovn-encap-ip="10.89.0.5", ovn-encap-type=geneve, ovn-is-interconn="true", ovn-monitor-all="true", ovn-ofctrl-wait-before-clear="0", ovn-openflow-probe-interval="180", ovn-remote="unix:/var/run/ovn/ovnsb_db.sock", ovn-remote-probe-interval="100000", ovn-set-local-ip="true", rundir="/var/run/openvswitch", system-id="02e441c7-2c53-493e-94d1-9c0bc14359b0"}
iface_types         : [afxdp, afxdp-nonpmd, bareudp, erspan, geneve, gre, gtpu, internal, ip6erspan, ip6gre, lisp, patch, srv6, stt, system, tap, vxlan]
manager_options     : []
next_cfg            : 45
other_config        : {bundle-idle-timeout="180", ovn-chassis-idx-02e441c7-2c53-493e-94d1-9c0bc14359b0="", vlan-limit="0"}
ovs_version         : "3.2.2"
ssl                 : []
statistics          : {}
system_type         : fedora
system_version      : "39"
```

We'll still see multiple patch ports (one per network) while using a single mapping - localnet `tenantblue` to bridge `breth0`. 

**- Description for the changelog**

Introduce a localnet network name alias, which allows OVN bridge mapping re-use.

### CODE REVIEW

File: go-controller/pkg/cni/types/types.go
Comment (maiqueb) [inline go-controller/pkg/cni/types/types.go:53]: @trozet we should agree on a name to put on the configuration to allow the users to re-use an existing mapping.

Maybe this is too low level / detail. How about directly `bridgeMapping` ?

File: test/e2e/localnet-underlay.go
Comment (maiqueb) [inline test/e2e/localnet-underlay.go:167]: maybe it wouldn't hurt to have this function idempotent ... 

It's really not needed right now, since we only use it from a single place, but could be helpful going forward.

File: go-controller/pkg/cni/types/types.go
Comment (phoracek) [reply go-controller/pkg/cni/types/types.go:53]: `bridgeMapping` would make sense to me. The user flow looks something like this:

1. Admin uses `NNCP` to configure `bridge-mappings`
```
    ovn:
      bridge-mappings:
      - localnet: localnet1 
        bridge: br-ex 
        state: present 
```
2. Admin proceeds to create a `NAD` where they have to reference `localnet1` from the example above. For me, the clearest hint to what I should be setting in the `NAD` would be if it asked for `bridgeMapping`. Another option would be `localnet`, but I like it less since we already use the term for the topology

Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go:53]: > The part that you actually put into the NAD is the physical network name

Good point; that is **really** what I am using - the value for the localnet port `network_name` option.

I am OK with `PhysicalNetworkName`. 

@phoracek thoughts ?

Comment (phoracek) [reply go-controller/pkg/cni/types/types.go:53]: Nice, that works for me

Comment (jcaamano) [reply go-controller/pkg/cni/types/types.go:53]: I am ok with the name itself. But I have some other thoughts.

While the network name is unique per network, this new field is no longer unique per network, same name can be used by multiple networks and can lead to invalid configurations. So what if there are multiple networks defined with different name but the same physical network and vlan ID? Suddenly we have to care about something that would otherwise be someone else's issue. Are we not going to care about it and let some higher orchestration layer UX take care of validating it?

From another perspective, how does OpenStack resolve this same use case? Do they have a single bridge mapping or multiple? If multiple, how does their workflow compare to ours, or what is in our workflow that makes it so cumbersome? Are we missing some orchestration UX that would automatically handle the bridge mapping to make this a non-issue? How would this feel to someone coming from OpenStack?

Both perspectives lead me to the same place: seems like a higher orchestration layer should take care of something extra. If this orchestration layer were in place, what makes the former option better than the latter?

Comment (phoracek) [reply go-controller/pkg/cni/types/types.go:53]: > So what if there are multiple networks defined with different name but the same physical network and vlan ID? 

That is completely valid scenario. Often localnet Pods won't be the only members of the VLAN. We will also have other non-Kubernetes services running there. So it is completely fine if we have NADD A and NAD B working in the same VLAN. If that is what you were concerned about.

But perhaps I did not understood what is the issue about orchestration.

Comment (jcaamano) [reply go-controller/pkg/cni/types/types.go:53]: > That is completely valid scenario. Often localnet Pods won't be the only members of the VLAN. We will also have other non-Kubernetes services running there. So it is completely fine if we have NADD A and NAD B working in the same VLAN. If that is what you were concerned about.

Well I guess that is true as long as there are no overlapping IPs. So my point is, it might be fine on some cases but not in others. And now knowing if it is fine or not can become our issue.

Comment (phoracek) [reply go-controller/pkg/cni/types/types.go:53]: I see, gotcha.

The way I'm looking at localnet is "you can use this to connect VMs to network managed by YOU". For VMs, we would typically rely on DHCP and don't use OVN IPAM. I understand that with Pods IPAM would be probably more common - there we should make it clear that it is admin's responsibility to make them non-overlapping.

I believe that secondary localnet UDN will make this easier. Localnet will be configurable through CUDN only, where you have the namespace selector. So I imagine people would typically create a single CUDN for a VLAN, matching many namespaces that should have access.

Now when I think about this - introducing this new "alias" is not making this issue worse, is it? I can have two mappings on a single bridge for two different networks with overlapping IPAM...

Localnet is dangerous by definition, it should be used by admins who know what they are doing. They have to make sure there is no collision with non-Kubernetes members of the network, and between Kubernetes members.

Comment (jcaamano) [reply go-controller/pkg/cni/types/types.go:53]: One of the first use cases would be to use `"PhysicalNetworkName": "physnet"`, right? That wouldn't be possible with the CUDN as is? Should this be a field in the CUDN as well? Otherwise, does this new field in the NAD make sense if the CUDN is expected to be used? I expect the CUDN to do the required validations in this case. 

In any case this answers my first question: this is a new concern for us and we plan to not really take care of it and leave it up to the admin. We know how that works for us however, we do end up dealing with it one way or another :stuck_out_tongue_winking_eye: 

> Now when I think about this - introducing this new "alias" is not making this issue worse, is it? I can have two mappings on a single bridge for two different networks with overlapping IPAM...

Right, but that wasn't ovn-k's problem because ovn-k was not handling this inter-connect between networks. Now it would be via this new Netconf API definition so we do need a clear statement on what we plan to do about it in ovn-k.

Comment (phoracek) [reply go-controller/pkg/cni/types/types.go:53]: Yeah, sharing the default network, with or without VLAN, would be very common.

Not quite sure what you mean about CUDN there. It is possible to share `physnet` using today's NAD, as long as you don't use any VLAN (@maiqueb keep me honest here). With PhysicalNetworkName the VLAN limitation would be gone. It should be also possible to reference `physnet` as any other PhysicalNetworkName from CUDN, once it gets `localnet` topology option.

> and we plan to not really take care of it and leave it up to the admin. 

+++, i like it. We should be aware of the use-case. And we should draw a clear line of "here we help you manage it, there you have to manage it yourself"

> wasn't ovn-k's problem because ovn-k was not handling this inter-connect between networks

We do support localnet since 4.14. We allowed users to connect to br-ex native network https://docs.openshift.com/container-platform/4.14/networking/multiple_networks/configuring-additional-network.html#configuring-additional-network_ovn-kubernetes-configuration-for-a-localnet-topology.  However, +1 on providing a clear statement and guidance to admins and even within OpenShift

Thank you Jaime!

Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go:53]: > > Not quite sure what you mean about CUDN there.
> 
> Hmm I guess I wasn't quite sure what you were referring to with CUDN either, you used the term first 😛
> 
> I was thinking on the UDN CRD, but we don't support localnet for primary networks and I am not sure we support the namespace selector for secondary networks. So not sure a consideration for the UDN CRD applies at all here.

We want to support the namespace selector for the secondary localnet UDNs for sure.

> 
> > We do support localnet since 4.14. We allowed users to connect to br-ex native network
> 
> What I mean is that effectively, inter-connecting two different ovn-k localnet networks was not possible in ovn-k. You would have to do that through the infra network which is not ovn-k responsibility. Now with this `PhysicalNetworkName`, kinda makes that interconnect possible (at least semantically in the ovn-k netconf API) through ovn-k so it would be reasonable to think that any sanity checks for that fall under ovn-k responsibility.

I don't follow: you can have multiple networks in the same OVS bridge *today* with the localnet feature. Each network will use a different patch port. 

All this PR does is allow these multiple networks to reach the bridge (still using different patch ports) using localnet LSPs whose network name is the same. 

The admin is still responsible to ensure the physical network configuration makes sense. 

Maybe a couple of examples of whatever sanity checks you have in mind would make this clearer for me.

Comment (jcaamano) [reply go-controller/pkg/cni/types/types.go:53]: > I don't follow: you can have multiple networks in the same OVS bridge _today_ with the localnet feature. Each network will use a different patch port.

I am not explaining myself very well. I know you can do that today. I am not saying you can't.

But doing so today requires two bridge mappings `physnet:br-ex,localnet:br-ex`. OVN-K does not care, is not aware nor handles the bridge mappings, so if this leads to any problem it is not OVN-K responsibility. `Something else` is setting the bridge mappings and that `something else` is responsible.

Now with this new attribute `PhysicalNetworkName` as part of the OVN-K netconf API and set to `physnet` you only need the default bridge mapping `physnet:br-ex` to have the equivalent functionality as before. In the general case this attribute makes possible to map multiple networks to the same bridge, **through OVN-K API**, introducing a gray area of responsibility whereas before it was clearly not OVN-K responsibility because OVN-K was simply not aware.

One example of the checks we could potentially make:
Check that all the localnets with the same `PhysicalNetworkName` and VlanID don't have overlapping IPs.

> We want to support the namespace selector for the secondary localnet UDNs for sure.

So the plan is also to have this field in the CRD as well?

Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go:53]: > > We want to support the namespace selector for the secondary localnet UDNs for sure.

> So the plan is also to have this field in the CRD as well?

Yes.

> Now with this new attribute PhysicalNetworkName as part of the OVN-K netconf API and set to physnet you only need the default bridge mapping physnet:br-ex to have the equivalent functionality as before. In the general case this attribute makes possible to map multiple networks to the same bridge, through OVN-K API, introducing a gray area of responsibility whereas before it was clearly not OVN-K responsibility because OVN-K was simply not aware.

I understand better what you mean. 

IMHO, we can simply call out in the docs this is *not* a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.

> One example of the checks we could potentially make:
Check that all the localnets with the same PhysicalNetworkName and VlanID don't have overlapping IPs.

Thanks for the example; I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined. I.e. ovn-k has no clue of the IP assignments of the workloads in it.

While saying that, I do see the potential for an higher orchestration layer you've called out in previous comments.

Comment (jcaamano) [reply go-controller/pkg/cni/types/types.go:53]: > IMHO, we can simply call out in the docs this is _not_ a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.

That works for the NAD but I don't think we can do that for the CRD because validation is one of its main purposes.

> I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined.

For the ones that have it defined I guess. We have the responsibility to think about all use cases not just virt.

Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go:53]: > > IMHO, we can simply call out in the docs this is _not_ a gray area of responsibility, and that OVN-K is in no way responsible for validating the physical network configuration - the admin is.
> 
> That works for the NAD but I don't think we can do that for the CRD because validation is one of its main purposes.

Sure, let's defer validation to the cUDN CRD then. That makes sense to me.

> 
> > I just don't see how can we perform or enforce such validation: localnet networks (at least for virt use cases) are commonly known for not having subnets defined.
> 
> For the ones that have it defined I guess. We have the responsibility to think about all use cases not just virt.

I'm OK with best effort - i.e. validating whenever possible.

Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go]: Renamed the attribute to `PhysicalNetworkName`.

Please take another look @trozet

**Review by kyrtapz — COMMENTED** [review]
The code change looks good to me.

I think we should document the potential issues and support limitation for NADs with the same `physicalNetworkName` in our upstream docs. While doing that it might be worth it to add a sentence or two about the happy case too :)
Lets also agree on how to move forward one the additional validation for the localnet CRD and create something to track the work so it's not lost to time.

File: test/e2e/multihoming.go
Comment (maiqueb) [reply test/e2e/multihoming.go:938]: it doesn't introduce; the dependency is already there: https://github.com/ovn-org/ovn-kubernetes/blob/08a965003b2b3da204278a7161850167c0c559d0/test/e2e/multihoming.go#L697

I'm just extending it.

I agree it would be a good follow up to remove these dependencies altogether and use a golang http server.

File: test/e2e/localnet-underlay.go
Comment (maiqueb) [reply test/e2e/localnet-underlay.go]: Done.

File: go-controller/pkg/cni/types/types.go
Comment (maiqueb) [reply go-controller/pkg/cni/types/types.go:59]: Done.

**Comment by maiqueb** [comment]
> The code change looks good to me.
> 
> I think we should document the potential issues and support limitation for NADs with the same `physicalNetworkName` in our upstream docs. While doing that it might be worth it to add a sentence or two about the happy case too :) Lets also agree on how to move forward one the additional validation for the localnet CRD and create something to track the work so it's not lost to time.

IMHO, all validation should happen at the localnet CRD level: 
- we should allow the MTU attribute to be mutable
- we should allow the list of VLANs on the localnet to be mutable (at conceptual level - I do not actually mean an array in the localnet CRD-to-be)
- do we want the CIDR attribute to be mutable ? (I don't think there's a use case for that, but all in all, whatever argument for updating the MTU can be made for updating the CIDR ...)

Once a parameter on the network (like MTU / CIDR) is updated:
- the workloads connected to the "old" network would enter an unspecified state, where they may (or may not) have access to other workloads created with the "old" attributes.
- the `network` would enter an `OutOfSync` state. This would signal the network's behavior will be unspecified until it successfully reconciles. Once all workloads connected to the network have consumed the **updated** configuration, the network is back in sync.
- the workloads created on the network with the **new** configuration will be able to communicate between themselves.

Do you agree @kyrtapz ?

**Review by kyrtapz — APPROVED** [review]

**Review by kyrtapz — DISMISSED** [review]

File: docs/features/multiple-networks/multi-homing.md
Comment (ricky-rav) [inline docs/features/multiple-networks/multi-homing.md:206]: Did you want to show an example with `physicalNetworkName` here, as you did in this PR description? Otherwise I didn't get the purpose of this section :)

Comment (maiqueb) [reply docs/features/multiple-networks/multi-homing.md:206]: Really good catch; I pasted a couple of outdated examples I had which used the old attribute name.

Now it should be in sync !

Thanks again.

File: docs/features/multiple-networks/multi-homing.md
Comment (jcaamano) [inline docs/features/multiple-networks/multi-homing.md:174]: I am assuming this is specific to localnet. Should the doc be more clear that this is being ignored for other types?

What is the plan though for L2, L3 primaries that will also be connected to the outside world? Wouldn't we want parity?

Comment (maiqueb) [reply docs/features/multiple-networks/multi-homing.md:174]: > I am assuming this is specific to localnet. Should the doc be more clear that this is being ignored for other types?

It is specific to localnet - it is *exclusively* defined in the localnet network configuration reference [section](https://github.com/ovn-org/ovn-kubernetes/blob/1c92c4d39cb1fb2b9cb22c292f047db52e570f2e/docs/features/multiple-networks/multi-homing.md#network-configuration-reference-2). 

I don't know how to make it clearer than that. I can write there that it only applies to localnet, but again, this section **is** for localnet only. 

The same comment would apply to `vlanID`.

> What is the plan though for L2, L3 primaries that will also be connected to the outside world? Wouldn't we want parity?

I think L2 / L3 primaries have a lot more batteries included and are a lot more "opinionated" than secondary networks. 

We *may* want to have parity in the future (i.e. a configurable mapping to the physical network for the primary network), but right now, nobody's asking for it. I do think it is a reasonable ask though.

Comment (jcaamano) [reply docs/features/multiple-networks/multi-homing.md:174]: Do the L2 / L3 primaries ingress/egress through the same localnet port as the default network?

Comment (maiqueb) [reply docs/features/multiple-networks/multi-homing.md:174]: > Do the L2 / L3 primaries ingress/egress through the same localnet port as the default network?

No.

Comment (jcaamano) [reply docs/features/multiple-networks/multi-homing.md:174]: I am just trying to guess which bridge mapping they are using. Do they hardcode `physnet` as network name then?

All this was more clear when there were just secondary networks and localnet was the only topology connected with the outside world. We could say this document is specifically for secondary networks. An explicit acknowledgement that it does not cover primary networks could also help. It's confusing because there is no equivalent doc for primaries, no doc covering the `Role` attribute of the NAD, and the last section on this doc covering a topic on primaries further confuses things. And to add to the confusion, you go to the code and see that `IsSecondary` returns true for primaries as well.

So if I see `physicalNetworkName` here, even if it is not documented in other sections, I would ask myself `Why wouldn't it work for other topologies for which this would make sense?` And then I would go around looking for other hints and probably be misguided. 

I am just saying this as an outsider that has not been deeply involved in UDN. I had that question and a clarification that explicitly states `This only applies to localnet secondary networks` would have helped me at a user level. But if you think it is fine then lets go with it.

Comment (maiqueb) [reply docs/features/multiple-networks/multi-homing.md:174]: I acknowledge all those limitations - but, to me, what's really missing is a document explaining primary UDN features.

FWIW, I think we should also do a major overhaul of the multi-homing docs after primary UDN lands. 

@RamLavi started something like that, but with a different scope in https://github.com/ovn-org/ovn-kubernetes/pull/4537

Comment (jcaamano) [reply docs/features/multiple-networks/multi-homing.md:174]: Do we have some way to track that future effort?

**Comment by jcaamano** [comment]
Waiting for all comments to be addressed. I think this one is pending: https://github.com/ovn-org/ovn-kubernetes/pull/4320#discussion_r1816398459

File: test/e2e/localnet-underlay.go
Comment (maiqueb) [reply test/e2e/localnet-underlay.go:167]: Done.

File: docs/features/multiple-networks/multi-homing.md
Comment (maiqueb) [reply docs/features/multiple-networks/multi-homing.md:174]: Do you mean the primary UDN docs ? If so, yes: https://issues.redhat.com/browse/SDN-5425

**Comment by maiqueb** [comment]
[Rebased](https://github.com/ovn-org/ovn-kubernetes/compare/1b83247fa4ae0105d8bcbb6009ca02701e838ea1..ae6ccc0123f743b926ceb3f042cca86579aedb07) the code.

**Review by jcaamano — APPROVED** [review]

File: test/e2e/multihoming.go
Comment (flavio-fernandes) [inline test/e2e/multihoming.go:906]: NOTE: this function "ovsPods" assumes that ovs is being ran as a pod, which is a problem for clusters that have ovs as part od systemctl in the node.

**Pull request merged**
